From da3a9b351e2f51469d3f9f83dde00f33aacf39b0 Mon Sep 17 00:00:00 2001
From: Yongxin Liu <yongxin.liu@windriver.com>
Date: Thu, 13 Jun 2024 10:26:44 +0000
Subject: [PATCH 13/24] rdk/snr: klm/cpk/ice_ptp.c: convert .adjfreq to
 .adjfine

kernel commit 75ab70ec5cef ("ptp: remove the .adjfreq interface
function").

Upstream-Status: Inappropriate [RDK is released in tarball only]

Signed-off-by: Yongxin Liu <yongxin.liu@windriver.com>
---
 klm/cpk/ice_ptp.c | 31 +++++++++++++++++++++++++++----
 1 file changed, 27 insertions(+), 4 deletions(-)

diff --git a/klm/cpk/ice_ptp.c b/klm/cpk/ice_ptp.c
index 926225a..2c0b504 100644
--- a/klm/cpk/ice_ptp.c
+++ b/klm/cpk/ice_ptp.c
@@ -2968,6 +2968,9 @@ exit:
 	return status;
 }
 
+#define ICE_ETH56G_NOMINAL_INCVAL        0x140000000ULL
+#define LOCKED_INCVAL_E822               0x100000000ULL
+
 /**
  * ice_ptp_adjfreq - Adjust the frequency of the clock
  * @ptp: The PTP clock structure
@@ -2976,11 +2979,13 @@ exit:
  * Adjust the frequency of the clock by the indicated parts per billion from the
  * base frequency.
  */
-static int ice_ptp_adjfreq(struct ptp_clock_info *ptp, s32 ppb)
+static int ice_ptp_adjfine(struct ptp_clock_info *ptp, long scaled_ppm)
 {
 	struct ice_vsi *vsi = container_of(ptp, struct ice_vsi, ptp_caps);
 	struct ice_pf *pf = vsi->back;
-	enum ice_status status;
+	u64 incval, diff;
+	int neg_adj = 0;
+	int status;
 
 	if (vsi->mstr_tmr_mode == ICE_MSTR_TMR_MODE_LOCKED) {
 		dev_err(ice_pf_to_dev(pf),
@@ -2988,7 +2993,25 @@ static int ice_ptp_adjfreq(struct ptp_clock_info *ptp, s32 ppb)
 		return -EIO;
 	}
 
-	status = ice_ptp_set_increment(pf, ppb);
+	if (vsi->mstr_tmr_mode == ICE_MSTR_TMR_MODE_NANOSECONDS)
+		incval = ICE_ETH56G_NOMINAL_INCVAL;
+	else
+		incval = LOCKED_INCVAL_E822;
+
+	if (scaled_ppm < 0) {
+		neg_adj = 1;
+		scaled_ppm = -scaled_ppm;
+	}
+
+	diff = mul_u64_u64_div_u64(incval, (u64)scaled_ppm,
+			1000000ULL << 16);
+
+	if (neg_adj)
+		incval -= diff;
+	else
+		incval += diff;
+
+	status = ice_ptp_set_incval(pf, incval);
 
 	if (status) {
 		dev_err(ice_pf_to_dev(pf), "PTP failed in adj freq %s\n",
@@ -4395,7 +4418,7 @@ static long ice_ptp_create_clock(struct ice_pf *pf)
 	vsi->ptp_caps.n_ext_ts = 0;
 	vsi->ptp_caps.pps = 1;
 	vsi->ptp_caps.n_per_out = 2;
-	vsi->ptp_caps.adjfreq = ice_ptp_adjfreq;
+	vsi->ptp_caps.adjfine = ice_ptp_adjfine;
 	vsi->ptp_caps.adjtime = ice_ptp_adjtime;
 #ifdef HAVE_PTP_CLOCK_INFO_GETTIME64
 	vsi->ptp_caps.gettime64 = ice_ptp_gettime;
-- 
2.39.3

